flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java

```diff
@@ -551,18 +579,47 @@ private boolean isSelfIgnore(final StateTransition<T> edge) {
 								nextVersion,
 								startTimestamp);
 					}
+
+					switch (skipStrategy.getStrategy()) {
+						case SKIP_PAST_LAST_EVENT:
+							if (nextState.isFinal()) {
+								resultingComputationStates.add(createStartComputationState(computationState, event));
+							}
+							break;
+						case SKIP_TO_FIRST:
+							if (nextState.getName().equals(skipStrategy.getPatternName()) &&
+								!nextState.getName().equals(currentState.getName())) {
+								ComputationState<T> startComputationState = createStartComputationState(computationState, event);
+								if (callLevel > 0) {
```

Comment:
Because we need to detect whether there is an infinite loop. I use the callLevel to track it here.